flips/20220203-capability-controllers.md

```diff
@@ -0,0 +1,341 @@
+
+```cadence
+let publicAccount = getAccount(issuerAddress)
+let countCap = publicAccount.borrowCapability<&{HasCount}>(/public/hasCount)!
+countRef.count
+```
+
+### Capability Controllers (CapCons)
+
+Each Capability would have an associated CapCon that would be created together with the Capability (Capabilities and Capability Controllers are in a 1 to 1 relation). The data associated with CapCons would be stored in arrays, so that each storage path on an account has an array of CapCons of Capabilities issued from that storage path.
+
+CapCons are a non-storable object.
+
+Capabilities also have an address field and a target field. The target field is the storage path of the targeted object. Creating a capability from a capability should be illegal.
+
+The definition of the `CapabilityController` .
+
+```cadence
+// CapabilityController can be retrieved via:
+// - authAccount.getControllers(path: StoragePath): [CapabilityController]
+// - authAccount.getController(capabilityId: UInt64): CapabilityController?
+struct CapabilityController {
+   // The block height when the capability was created.
+   let issueHeight: UInt64
+   // The id of the related capability
+   // This is the UUiD of the created capability
+   let capabilityId: UInt64
+
+   // Is the capability revoked.
+   fun isRevoked(): Bool
+   // The target storage path.
+   fun target(): StoragePath
+   // Revoke the capability.
+   // Returns true if successfully revoked.
+   fun revoke(): Bool
+   // Restore the capability.
+   // Returns true if successfully restored.
+   fun restore(): Bool
+   // Retarget the capability.
+   // Returns if successfully restored.
+   // This would move the CapCon from one CapCon array to another
+   fun retarget(target: StoragePath): Bool
+}
+```
+
+The auth account would get new methods to get CapCons in order to manage capabilities:
+
+```cadence
+// Get all capability controllers for capabilities that target this storage path
+fun getControllers(path: StoragePath): [CapabilityController]
+// Get capability controller for capability with the specified id
+// If the id does not reference an existing capability
+// or the capability does not target a storage path on this address, return nil
+fun getController(capabilityId: UInt64): CapabilityController?
+```
+
+Some methods would be removed from the AuthAccount object as they are no longer needed:
+
+```cadence
+fun link<T: &Any>(_ newCapabilityPath: CapabilityPath, target: Path): Capability<T>?
+fun getLinkTarget(_ path: CapabilityPath): Path?
+fun unlink(_ path: CapabilityPath)
+```
+
+The method `getCapability` would be renamed to `issueCapability `to reflect the fact that a new capability is created every time.
```

Comment:
As noted by @aishairzay, we maybe want to consider adding a function that allows issuing a capability by providing a run-time type